lib/vrl/parser/src/template_string.rs

```diff
@@ -0,0 +1,76 @@
+use diagnostic::Span;
+use std::fmt;
+
+use crate::ast::{Expr, Ident, Literal::RawString, Node, Op, Opcode};
+
+#[derive(Clone, PartialEq, Eq, Ord, PartialOrd, Debug, Hash)]
+pub struct TemplateString(pub Vec<StringSegment>);
+
+impl TemplateString {
+    /// Rewrites the ast for the template string to be a series of string concatenations
+    pub fn rewrite(&self) -> Expr {
+        self.0
+            .iter()
+            .map(|node| -> Expr {
+                match node {
+                    StringSegment::Literal(s) => {
+                        Expr::Literal(Node::new(diagnostic::Span::default(), RawString(s.clone())))
+                    }
+                    StringSegment::Template(s, span) => {
+                        Expr::Variable(Node::new(*span, Ident::new(s)))
+                    }
```

Comment:
It might be worth adding a `From<StringSegment> for Exp {}` impl, so this while `map` can be `map(Into::into)`

Reply:
I could, but it is a reference at this point so it would be `From<&StringSegment>` which is not quite as idiomatic..

Reply:
I don't see why it isn't. It doesn't appear we need to clone anything, so that seems fine by me?

Reply:
It has all changed now anyway as I am passing the span information through too.